nni/algorithms/compression/pytorch/pruning/transformer_pruning_head_masker.py

```diff
@@ -0,0 +1,434 @@
+                n_selected += 1
+                # update pruned_heads in pruner (mainly for iterative pruning)
+                self.pruner.pruned_heads[weight_group[0].group_idx].add(head_idx)
+            if n_selected == num_prune:
+                break
+
+        return self._get_layer_masks_from_head_mask(weight_group, head_mask_bool)
+
+    def get_head_importance_scores(self, weight_group):
+        """
+        Calculate the importance score for each head.
+        Parameters
+        ----------
+        weight_group: list
+            list of a group of weights for an attention layer
+
+        Returns
+        -------
+        importance_scores: tensor
+            Tensor that indicates the importance of each head
+        """
+        raise NotImplementedError('{} get_channel_sum is not implemented'.format(self.__class__.__name__))
+
+
+class L1WeightHeadMasker(AttentionHeadMasker):
+    """
+    A structured pruning algorithm that prunes the heads weight smallest weight magnitude for the query, head,
+    and key projection matrices. L1 norm is used for magnitude calculation. Note that in this implementation, weight
+    norms of q_proj, k_proj, v_proj from each head are summed as the final importance score for the head.
+    """
+    def get_head_importance_scores(self, weight_group):
+        q_proj, k_proj, v_proj, _ = weight_group
+
+        n_heads = q_proj.module.weight.size()[0] // self.head_hidden_dim
+        query_proj_weights = q_proj.module.weight.data.view([n_heads, -1])
+        key_proj_weights = k_proj.module.weight.data.view([n_heads, -1])
+        value_proj_weights = v_proj.module.weight.data.view([n_heads, -1])
+
+        query_norm_avg = torch.sum(torch.abs(query_proj_weights), -1)
+        key_norm_avg = torch.sum(torch.abs(key_proj_weights), -1)
+        value_norm_avg = torch.sum(torch.abs(value_proj_weights), -1)
+
+        return ((query_norm_avg + key_norm_avg + value_norm_avg) / 3).detach()
+
+    def get_mask(self, num_prune, weight_group, **kwargs):
+        return self.get_mask_by_importance_ranking(num_prune, weight_group)
+
+
+class L2WeightHeadMasker(AttentionHeadMasker):
+    """
+    A structured pruning algorithm that prunes the heads weight smallest weight magnitude for the query, head,
+    and key projection matrices. L2 norm is used for magnitude calculation. Note that in this implementation, weight
+    norms of q_proj, k_proj, v_proj from each head are summed as the final importance score for the head.
+    """
+    def get_head_importance_scores(self, weight_group):
+        q_proj, k_proj, v_proj, _ = weight_group
+
+        n_heads = q_proj.module.weight.size()[0] // self.head_hidden_dim
+        query_proj_weights = q_proj.module.weight.data.view([n_heads, -1])
+        key_proj_weights = k_proj.module.weight.data.view([n_heads, -1])
+        value_proj_weights = v_proj.module.weight.data.view([n_heads, -1])
+
+        query_norm_avg = torch.sum(query_proj_weights ** 2, -1)
```

Comment:
why we don't use `torch.norm()`? and missing `torch.sqrt()`?

Reply:
`torch.norm()` is cleaner. I will fix this.